### ISSUE TITLE
Redundant Enablement

### ISSUE THREAD

**Issue by micahalcorn** [MetaMask/metamask-extension#5825]
**Describe the bug**
If a user has previously approved a DApp's 1102 request, an unnecessary, redundant request will be displayed if the `ethereum.enable()` method is called while the user is signed out.

**To Reproduce**
1. Call `ethereum.enable()`
2. Approve the request
3. Sign out of MetaMask
4. Call `ethereum.enable()`
5. Sign in to MetaMask
6. Reject the new request at your peril

**Expected behavior**
If `ethereum.enable()` is called for a user who has already approved the DApp, additional approval dialogs will not be displayed.

**Screenshots**
Demonstration with CryptoKitties:

https://www.useloom.com/share/b4a6a58d18df41fd97513840a5784d3e

**Browser details (please complete the following information):**
 - OS X
 - Chrome
 - 5.0.3

**Additional context**
This is demonstrated using CryptoKitties to show that it happens with a respected project in the wild. It can be consistently reproduced locally. cc @bitpshr (as described [here](https://github.com/MetaMask/metamask-extension/issues/5404#issuecomment-441387813))

**Comment by bitpshr** [comment]
Hi @micahalcorn. Thanks for taking the time to file this issue. This is intended behavior (related to [#5679](https://github.com/MetaMask/metamask-extension/issues/5679#issuecomment-436037424) and [#5697](https://github.com/MetaMask/metamask-extension/issues/5697#issuecomment-437029430)): if a dapp calls `ethereum.enable()` and MetaMask is locked, an approval popup will show that allows the user to both log in and approve the request regardless of cached approval. We could technically simplify this and only require the user to log in if a cached approval exists, but the UX of closing the popup immediately after login is jarring. We're investigating this internally, but for now, the behavior described is intended. Also, it's important to note that approval caching was always intended to be a user-facing feature that dapps should not rely on. In the future, we'll allow users to turn caching off altogether if they desire, so dapps should avoid UX flows that rely on this feature.

Don't hesitate to let us know if you encounter any additional issues.

**Comment by bitpshr** [comment]
Hi @micahalcorn, thanks for the additional input.

Users will only be required to approve a previously-enabled dapp if their wallet is locked. This can happen for two primary reasons: if the user explicitly locks MetaMask, or if the user visits a dapp after restarting their browser without first unlocking MetaMask. Sometimes this will result in per-session approval, sometimes it won't; this purely depends on the state of a given user's MetaMask. Again, caching approvals at all is a MetaMask-specific feature and may or may not be implemented in other dapp browsers. This feature will also be able to be toggled "off" by users in the future.

While we strive to cache approvals as aggressively as possible, we're not confident that the UX of immediately closing an approval popup after the user enters their password with no additional context (like a subsequent approval screen) is fluid or safe. This is the whole reason we ignore cached approvals and require that users explicitly acknowledge what they are entering their password for. We opted to go with a slightly-more obtrusive experience (or a less-aggressively-cached experience) over a potentially-unsafe experience (for now, at least.)

Thanks again for the additional input. We're actively investigating possible solutions, and your added insight really helps. Closing, but don't hesitate to open new issues if you find anything else out of place.
